### PR TITLE
feat: transcribe meetings with cal-ai

### LINKED ISSUE #11013: [CAL-2903] Transcribe and summarize video meetings using AI
### Is your proposal related to a problem?

I use Cal to schedule user interviews for [Telborg](https://telborg.com), then take handwritten notes during the video meeting and rewrite those notes in Slab. 

### Describe the solution you'd like

I'd love a workflow where I can:
1. request consent to record automatically when meeting starts
2. transcribe the conversation using a speech-to-text model such as [Whisper](https://platform.openai.com/docs/models/whisper)
3. Summarise the conversation using GPT 3.5/4 /other LLM
4. Send transcript + summary to Notion/[Slab](https://slab.com/)/Google Docs/other wiki provider
5. (Optional) send follow-up email to meeting attendees summarising the conversation

### Describe alternatives you've considered

I currently use Google Meet for the meetings. A Zapier workflow with access to Google Drive might be possible but I would have to
1. remember to ask for consent and enable recording
2. setup zapier + google drive to access recording
3. setup zapier + openai to transcribe and summarise
4. setup zapier  + notion/slab/google docs to store transcript and summary

The whole setup can be simplified by 
1. using [daily.co's transcription service](https://docs.daily.co/reference/rn-daily-js/instance-methods/start-transcription#main) - this has native support for [Whisper](https://developers.deepgram.com/docs/model#whisper) 
2. building Cal integrations for the wiki providers we want to support

### Additional context

I love Loom's AI features - using AI to transcribe the recording, generate a summary, generate a relevant title. And all of this happens nearly real-time.

Happy to work on this independently or with other contributors!

### Requirement/Document

<sub>[CAL-2903](https://linear.app/calcom/issue/CAL-2903/transcribe-and-summarize-video-meetings-using-ai)</sub>

### PR DESCRIPTION
**Pull request by aar2dee2**
## What does this PR do?

Fixes #11013 

## Requirement/Documentation

- If there is a requirement document, please, share it here.
- If there is ab UI/UX design document, please, share it here.

## Type of change

- [-] New feature (non-breaking change which adds functionality)

## How should this be tested?

- Are there environment variables that should be set?
- What are the minimal test data to have?
- What is expected (happy path) to have (input and output)?
- Any other important info that could help to test that PR

## Mandatory Tasks

- [ ] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

## Checklist

- I haven't checked if my PR needs changes to the documentation
- I haven't checked if my changes generate no new warnings
- I haven't added tests that prove my fix is effective or that my feature works
- I haven't checked if new and existing unit tests pass locally with my changes

### CODE REVIEW

File: packages/app-store/dailyvideo/lib/VideoApiAdapter.ts [hunk 1]
```diff
@@ -130,6 +137,8 @@ const DailyVideoApiAdapter = (): VideoApiAdapter => {
           enable_chat: true,
           exp: exp,
           enable_recording: "cloud",
+          enable_transcription: `deepgram:${process.env.DEEPGRAM_API_KEY}`,
+          permissions: { canAdmin: ["transcription"] },
```
Comment (PeerRich) [inline packages/app-store/dailyvideo/lib/VideoApiAdapter.ts:141]: we should also add it to https://github.com/calcom/cal.com/blob/main/apps/ai/.env.example

File: apps/web/pages/video/[uid].tsx [hunk 2]
```diff
@@ -51,6 +63,14 @@ export default function JoinCall(props: JoinCallPageProps) {
           supportiveText: "#FFF",
         },
       },
+      customTrayButtons: {
+        transcriptButton: {
+          iconPath: "https://unpkg.com/lucide-static@latest/icons/sparkles.svg", // TODO
+          iconPathDarkMode: "https://unpkg.com/lucide-static@latest/icons/sparkles.svg", //TODO
```
Comment (PeerRich) [inline apps/web/pages/video/[uid].tsx:69]: i think we can move these into our public/ folder

File: packages/app-store/cal-ai/DESCRIPTION.md [hunk 3]
```diff
@@ -31,4 +31,6 @@ items:
 
 
 ### Or, doing multiple things:
-- Cancel all my meetings tomorrow then book a quick chat with @joe at noon, a long meeting with @jane at 1 (both tomorrow). Also see if Toby is down to grab dinner this week (toby@gmail.com). Finally send me my updated agenda of everything I have coming up in the next couple days.
\ No newline at end of file
+- Cancel all my meetings tomorrow then book a quick chat with @joe at noon, a long meeting with @jane at 1 (both tomorrow). Also see if Toby is down to grab dinner this week (toby@gmail.com). Finally send me my updated agenda of everything I have coming up in the next couple days.
+You will need a [Deepgram](https://console.deepgram.com/) API Key to transcribe calls in the app.
+Go to `Settings -> Admin -> Apps interface` and add the `deepgram_api_key` for `Cal.ai`.
```
Comment (PeerRich) [inline packages/app-store/cal-ai/DESCRIPTION.md:36]: does every user need a Deepgram API Key? Can we not provide one for our users?

File: apps/web/pages/video/[uid].tsx (hunk 2 above)
Comment (aar2dee2) [reply apps/web/pages/video/[uid].tsx:69]: we need to use the url.
We can fix the css by adding styles for `#transcriptButton-controls`

File: packages/app-store/cal-ai/DESCRIPTION.md (hunk 3 above)
Comment (aar2dee2) [reply packages/app-store/cal-ai/DESCRIPTION.md:36]: You could, but then you probably need one of their plans depending on the usage?

File: apps/web/pages/video/[uid].tsx [hunk 4]
```diff
@@ -60,11 +80,66 @@ export default function JoinCall(props: JoinCallPageProps) {
       url: meetingUrl,
       ...(typeof meetingPassword === "string" && { token: meetingPassword }),
     });
+    //TODO - add handlers for `transcription-started` and `transcription-ended` @link https://docs.daily.co/reference/rn-daily-js/events/transcription-events
+    /** handling custom-button-click @link  https://docs.daily.co/reference/daily-js/events/meeting-events#custom-button-click */
+    const onCustomButtonClick = async (event?: DailyEventObjectCustomButtonClick | undefined) => {
+      console.log(`calAiCredential ${JSON.stringify(props.calAiCredential)}`);
+      const calAiKeys = await getAppKeysFromSlug("cal-ai");
```
Comment (emrysal) [inline apps/web/pages/video/[uid].tsx:87]: You can't do this clientside. - getAppKeysFromSlug is server only.

Comment (aar2dee2) [reply apps/web/pages/video/[uid].tsx:87]: oh, okay. I'll pass it in props. thanks!

File: apps/web/pages/video/[uid].tsx [hunk 5]
```diff
@@ -85,7 +160,24 @@ export default function JoinCall(props: JoinCallPageProps) {
...
+    //TODO - do we need anything here?
+  };
+  const handleTranscriptionStopped = async () => {
+    console.log("handleTranscriptionStopped is called");
+    if (!props.calAiCredential || !props.calAiCredential.key || !transcript.current) {
+      console.log("returning without calling the ai tool");
+      return;
+    }
+    //TODO
+    // send the transcript to the transcription meeting handler
```
Comment (emrysal) [inline apps/web/pages/video/[uid].tsx:173]: This relies on sendEmail; you'll need to do something similar to the above to get this to work through the backend, currently it's throwing an error it tries to send email from the browser.

```javascript
  fetch("/api/recorded-daily-video", {
      method: "POST",
      body: JSON.stringify(data),
    }).catch((err) => {
      console.log(err);
    });
    recordingId.current = null;
  };
```

Comment (aar2dee2) [reply apps/web/pages/video/[uid].tsx:173]: going to try this. thanks!

File: apps/ai/.env.example [hunk 6]
```diff
@@ -13,6 +13,7 @@ SENDER_DOMAIN=cal.ai
 PARSE_KEY=
 
 OPENAI_API_KEY=
+DEEPGRAM_API_KEY=
```
Comment (PeerRich) [inline apps/ai/.env.example:16]: sweet. added to vercel prod

Comment (PeerRich) [reply apps/ai/.env.example:16]: log in is in our password manager

**Review by zomars — CHANGES_REQUESTED** [review]
Thank you for your contribution 🙏 

I'm noticing some observation to address before feeling confidence to merge and ship:

- The "How should this be tested?" is not filled properly
- "I haven't added tests that prove my fix is effective or that my feature works" this one is also important
- Checks are currently failing as well


EDIT: 


These are the type errors:

image

File: apps/web/pages/video/[uid].tsx [hunk 7]
```diff
@@ -60,11 +79,64 @@ export default function JoinCall(props: JoinCallPageProps) {
       url: meetingUrl,
       ...(typeof meetingPassword === "string" && { token: meetingPassword }),
     });
+    //TODO - add handlers for `transcription-started` and `transcription-ended` @link https://docs.daily.co/reference/rn-daily-js/events/transcription-events
+    /** handling custom-button-click @link  https://docs.daily.co/reference/daily-js/events/meeting-events#custom-button-click */
+    const onCustomButtonClick = async (event?: DailyEventObjectCustomButtonClick | undefined) => {
+      console.log(`calAiCredential ${JSON.stringify(props.calAiCredential)}`);
+
+      if (!props.calAiCredential || !props.calAiCredential.key || !props.calAiCredential.key["apiKey"]) {
```
Comment (zomars) [inline apps/web/pages/video/[uid].tsx:87]: - This is causing a type error
- Do we want credentials keys on the frontend? This makes them visible to anyone inspecting the HTML code.

**Comment by aar2dee2** [comment]
We only want to check if cal-ai is installed. I can pass it as a Boolean
from props.

**Review by emrysal — CHANGES_REQUESTED** [review]
New information shows we can use an experimental API to achieve this much easier, normally wouldn't advocate for experimental APIs but as this is a background task using this new API it's got my vote.

**Comment by aar2dee2** [comment]
Here's a video showing the earlier approach using the [`startTranscription` API](https://docs.daily.co/reference/daily-js/instance-methods/start-transcription). I'm running into a `CORS` error. Would love help on fixing that.
[Loom](https://www.loom.com/share/5a00355918ea4e60b1b54d05ff227c8a?sid=fb040c1b-823c-413c-afe1-fc53c170c840)

If we use the [Batch Processor API](https://docs.daily.co/private/batch-processor/reference) there are the following considerations:
1. This approach generates transcripts from call recordings. Need to look into
- whether to offer call recordings on individual free plan or offer transcription only on the team plan
- There won’t be a separate `Cal.ai` button. Need to let the user know that they must record the call for transcription. (In Google Meet, when you click `Start Recording`, there’s a checkbox for enabling transcript below the button)

2. The Batch Process API removes identifying information so we do not get speaker names in the transcript. Daily team mentioned that at the moment they will not be able to support speaker names. Here's what a transcript looks like using this API
Screenshot 2023-10-24 at 11 47 40 PM

3. Need to setup crons to check is the `transcript` and `summarize` jobs have been completed

File: .env.example [hunk 8]
```diff
@@ -43,7 +43,7 @@ NEXT_PUBLIC_WEBAPP_URL='http://localhost:3000'
 NEXT_PUBLIC_WEBSITE_URL='http://localhost:3000'
 NEXT_PUBLIC_CONSOLE_URL='http://localhost:3004'
 NEXT_PUBLIC_EMBED_LIB_URL='http://localhost:3000/embed/embed.js'
-
+CAL_AI_PARSE_KEY=''
```
Comment (Udit-takkar) [inline .env.example:46]: @aar2dee2 what is CAL_AI_PARSE_KEY ?

Comment (Udit-takkar) [reply .env.example:46]: and what is CAL_AI_URL?

File: packages/app-store/dailyvideo/lib/VideoApiAdapter.ts [hunk 9]
```diff
@@ -120,28 +133,46 @@ const DailyVideoApiAdapter = (): VideoApiAdapter => {
...
+    };
+    /** permission to allow transcription using cal-ai @link https://docs.daily.co/reference/rn-daily-js/instance-methods/start-transcription#main */
     if (scalePlan === "true" && !!hasTeamPlan === true) {
+      if (checkCalAiInstalled) {
+        return {
+          privacy: "public",
+          properties: {
+            ...options,
+            enable_recording: "cloud",
+            enable_transcription: `deepgram:${process.env.DEEPGRAM_API_KEY}`,
```
Comment (Udit-takkar) [inline packages/app-store/dailyvideo/lib/VideoApiAdapter.ts:151]: I don't see any `enable_transcription` property in docs

**Pull request closed without merging**